### PR TITLE
Multi-request, split messages to multiple handlers.

### PR DESCRIPTION
**Pull request by nohwnd**
## Description
The idea here is that we extend the protocol on message to carry recipient as well, together with message type and content. (Similar to HTTP headers). This tells us which handler is waiting for the message. And on the wrapper side we dispatch the messages to the correct handler. 
This extension will be used only between vstest.console and the wrapper, because that is where we need to respond to multiple requests over the same channel, and will be considered only by the design mode client. Other handlers are safe to ignore it, as long as they play by the 
old rules of only using single request at the time.

This way we avoid changing every message that can be sent. And allow the vstest console wrapper to register multiple handlers that will receive the responses that are only meant for them. 

One "special" case is raw message. Raw messages are used to forward the text of the message over a hop without parsing it. (e.g. message from testhost goes via vstest.console to TestWindowStoreHost). This approach still supports this, it just wraps the message into another Message
and sends it as a non-raw message so it can have a recipient. Technically we could modify the recipient in the raw message, but that would require us to parse the message, which is what we want to avoid.

## Related issue
Kindly link any related issues. E.g. Fixes #xyz.

### CODE REVIEW

File: playground/TestPlatform.Playground/Program.cs [hunk 1]
```diff
@@ -49,42 +53,97 @@ static void Main(string[] args)
...
+            foreach (var (source, handler, launcher) in sources)
+            {
+                var options = new TestPlatformOptions { };
+                tasks2.Add(Task.Run(() =>
+                {
+                    r.DiscoverTests(new[] { source }, sourceSettings, options, handler);
+                    var tests = handler.DiscoveredTests;
+                    if (!tests.Any())
+                    {
+                        // REview: the second discovery can get cancelled too early. This is task that will probably be solved by fixing the cancellation in the translation layer sender and how it receives messages.
```
Comment (MarcoRossignoli) [inline playground/TestPlatform.Playground/Program.cs:79]:
```suggestion
                        // Review: the second discovery can get cancelled too early. This is task that will probably be solved by fixing the cancellation in the translation layer sender and how it receives messages.
```

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs [hunk 2]
```diff
@@ -4,22 +4,21 @@
...
-    /// Registers the discovery and test run events for design mode flow
-    /// </summary>
-    public class DesignModeTestEventsRegistrar : ITestDiscoveryEventsRegistrar, ITestRunEventsRegistrar
+    internal class IdentifiableDesignModeTestEventsRegistrar : ITestDiscoveryEventsRegistrar, ITestRunEventsRegistrar
     {
+        private string testRunId;
         private IDesignModeClient designModeClient;
 
-        public DesignModeTestEventsRegistrar(IDesignModeClient designModeClient)
+        public IdentifiableDesignModeTestEventsRegistrar(IDesignModeClient designModeClient, string testRunId)
```
Comment (MarcoRossignoli) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs:16]: I'd keep `recipient`  for clarity or do you have a reason to change the name?

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs [hunk 3]
```diff
@@ -28,6 +28,7 @@ namespace Microsoft.VisualStudio.TestPlatform.Client.DesignMode
     using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
 
     using CommunicationUtilitiesResources = CommunicationUtilities.Resources.Resources;
+    using System.Linq.Expressions;
```
Comment (MarcoRossignoli) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs:31]: cannot compile with eyes :) is it needed?

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs [hunk 4]
```diff
@@ -4,22 +4,21 @@
...
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
 
-    /// <summary>
-    /// Registers the discovery and test run events for design mode flow
-    /// </summary>
-    public class DesignModeTestEventsRegistrar : ITestDiscoveryEventsRegistrar, ITestRunEventsRegistrar
+    internal class IdentifiableDesignModeTestEventsRegistrar : ITestDiscoveryEventsRegistrar, ITestRunEventsRegistrar
     {
+        private string testRunId;
```
Comment (MarcoRossignoli) [inline src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs:13]: nit: `readonly`, also the other one maybe

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs [hunk 5]
```diff
@@ -484,36 +491,50 @@ public void InitializeExtensions(IEnumerable<string> pathToAdditionalExtensions)
                     TestPlatformOptions = options
                 };
 
+                var id = Id.Next().ToString();
+
+                if (!this.sessionEventHandlers.TryAdd(id, eventsHandler))
+                {
+                    throw new InvalidOperationException($"Adding session hancler under unique id {id} did not succeed, this should never happen.");
```
Comment (MarcoRossignoli) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:498]:
```suggestion
                    throw new InvalidOperationException($"Adding session handler under unique id {id} did not succeed, this should never happen.");
```

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs (hunk 2 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs:16]: I thought I need to rename this type, but I don't need to do that, so I will rename that back. And the testRunId was a previous design. So I will change that as well.

File: playground/TestPlatform.Playground/Program.cs (hunk 1 above)
Comment (nohwnd) [reply playground/TestPlatform.Playground/Program.cs:79]: Yeah, that task in the comment needs to be completed before merging.

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs (hunk 4 above)
Comment (Evangelink) [reply src/Microsoft.TestPlatform.Client/DesignMode/DesignModeTestEventsRegistrar.cs:13]: Happy to enable the related rule as warning after we merge the `.editorconfig` change if we agree on it.

File: src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs (hunk 3 above)
Comment (Evangelink) [reply src/Microsoft.TestPlatform.Client/DesignMode/DesignModeClient.cs:31]: Happy to enable the related rule as warning after we merge the `.editorconfig` change if we agree on it.

**Pull request closed without merging**